File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProduct = nums[0], minProduct = nums[0], best = nums[0], temp;

        /*
        Why track both maximum and minimum products ending at the current position?
        Because multiplying by a negative number can turn the smallest negative product into the largest positive product. 
        */

        for (int i = 1; i < nums.length; i++) {

          if (nums[i] < 0) {
            // if num is -ve, the previous minProduct can potentially become the new maxProduct.
            temp = maxProduct;
            maxProduct = minProduct;
            minProduct = temp;
          }

          maxProduct = Math.max(nums[i], nums[i] * maxProduct);
          minProduct = Math.min(nums[i], nums[i] * minProduct);
          best = Math.max(maxProduct, best);
        }
        return best;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = maxProduct = minProduct = nums[0]

        for x in nums[1:]:

          if x < 0: 
            maxProduct, minProduct = minProduct, maxProduct
          
          maxProduct = max(x, maxProduct * x)
          minProduct = min(x, minProduct * x)

          result = max(maxProduct, result)
        
        return result